server/db_accessor.cpp

```diff
@@ -4,59 +4,125 @@
         return false;
+    }
+
+    std::vector<char> hashed_password = std::get<1>(user_credential.at(0));
+    std::vector<char> salt = std::get<2>(user_credential.at(0));
+    bool logged_in = std::get<3>(user_credential.at(0));
 
     unsigned char buffer[HASH_SIZE] = {0};
     memcpy(buffer, password.data(), (std::min)(HASH_SIZE, password.size()));
     for (std::size_t i = 0; i < HASH_SIZE; i++) {
-        buffer[i] ^= iter->second.salt[i];
+        buffer[i] ^= salt[i];
     }
 
-    unsigned char digest[HASH_SIZE] = {0};
+    std::vector<char> digest(HASH_SIZE, 0);
     SHA256_CTX context;
     if (!SHA256_Init(&context))
         return false;
 
     if (!SHA256_Update(&context, (unsigned char *)buffer, HASH_SIZE))
         return false;
 
-    if (!SHA256_Final(digest, &context))
+    if (!SHA256_Final((unsigned char *)&digest[0], &context))
         return false;
 
-    if (memcmp(digest, iter->second.hashed_password, HASH_SIZE) == 0) {
-        if (iter->second.logged_in)
+    if (memcmp((unsigned char *)&digest[0], (unsigned char *)&hashed_password[0], HASH_SIZE) == 0) {
+        if (logged_in)
+            return false;
+
+        ozo::execute(conn, "UPDATE user_id SET logged_in = true WHERE username="_SQL + id , yield[ec]);
+        if (ec) {
+            BOOST_LOG_SEV(db_accessor::logger_, sev::error) << ec.message() << " | " << ozo::error_message(conn)
+                << " | " << ozo::get_error_context(conn);
             return false;
-        iter->second.logged_in = true;
+        };
         return true;
     }
-
     return false;
 }
 
-bool nibaserver::db_accessor::logout(const std::string &id) {
-    auto iter = db_.find(id);
-    if (iter == db_.end())
+bool nibaserver::db_accessor::logout(boost::asio::io_context &ioc, boost::asio::yield_context &yield, const std::string &id) {
+    ozo::error_code ec{};
+    ozo::rows_of<bool> logged_status;
+    auto conn = ozo::request(ozo::make_connector(conn_info, ioc),
+                             "SELECT logged_in FROM user_id WHERE username="_SQL + id,
+                             ozo::into(logged_status), yield[ec]);
+    if (ec) {
+        BOOST_LOG_SEV(db_accessor::logger_, sev::error) << ec.message() << " | " << ozo::error_message(conn)
+            << " | " << ozo::get_error_context(conn);
+        return false;
+    };
+
+    if (logged_status.size() == 0) {
+        BOOST_LOG_SEV(db_accessor::logger_, sev::warning) << "User " << id
+            << " is trying to logout, but the username is not in the database.";
         return false;
-    iter->second.logged_in = false;
+    }
+    
+    bool logged_in = std::get<0>(logged_status.at(0));
+    if (!logged_in) {
+        BOOST_LOG_SEV(db_accessor::logger_, sev::warning) << "User " << id
+            << " is trying to logout when the status is already logged out.";
+        return false;
+    }
+    
+    ozo::execute(conn, "UPDATE user_id SET logged_in = false WHERE username="_SQL + id , yield[ec]);
```

Comment:
I think we don't need to do any checking, just let the person log out... More sql = more opportunity to run into concurrency issues, especially without transactions and `select for update` etc.

Reply:
Ok, now logout() does not check whether the username exists in the database or the current status is logged_in. There is only one UPDATE to set logged_in = false. This change saves one SELECT query.